internal/server/recon/recognize.go

```diff
@@ -446,24 +455,67 @@ func getNumSpansForContainedIn(spans []*pb.TokenSpans_Span, startIdx int) int {
 	return 0
 }
 
-func combineContainedInSingle(
-	spans []*pb.TokenSpans_Span, startIdx, numSpans int) *pb.TokenSpans_Span {
+func combineTokenSpans(spans []*pb.TokenSpans_Span, startIdx, numSpans int) *pb.TokenSpans_Span {
 	startSpan := spans[startIdx]
-	endSpan := spans[startIdx+numSpans-1]
-
+	
 	res := &pb.TokenSpans_Span{Tokens: startSpan.Tokens}
 	for i := 1; i < numSpans; i++ {
 		res.Tokens = append(res.Tokens, spans[startIdx+i].GetTokens()...)
 	}
 
-	// This map is used to collect all the places for the combined span, with dedup.
+	return combineContainedInTokens(res, spans, startIdx, numSpans)
+}
+
+func getNextPlaceTokenSpan(spans []*pb.TokenSpans_Span, startIdx int) (*pb.TokenSpans_Span, int) {
+	for i := startIdx; i < len(spans); i++ {
+		if len(spans[i].GetPlaces()) > 0 {
+			return spans[i], i
+		}
+	}
+	return nil, -1
+}
+
+func findRecogPlaces(startSpan, nextSpan *pb.TokenSpans_Span) (map[string]*pb.RecogPlace, []string) {
 	dcidToRecogPlaces := map[string]*pb.RecogPlace{}
+	nextContainingPlaces := []string{}
 
 	for _, p1 := range startSpan.GetPlaces() {
 		for _, containingPlace := range p1.GetContainingPlaces() {
-			for _, p2 := range endSpan.GetPlaces() {
+			for _, p2 := range nextSpan.GetPlaces() {
 				if containingPlace == p2.GetDcid() {
 					dcidToRecogPlaces[p1.GetDcid()] = p1
+					nextContainingPlaces = p2.GetContainingPlaces()
```

Comment:
since we can get into this if statement multiple times, should we be appending to the nextContainingPlaces instead of replacing it? or have it be a set?

Reply:
Yes i think you're right. I just changed to appending.